server/main_server/src/main.cpp

```diff
@@ -0,0 +1,1471 @@
+
+    // NOLINTNEXTLINE(readability-function-cognitive-complexity)
+    auto migrate_and_launch_delay_server(bool _write_to_stdout,
+                                         bool _enable_test_mode,
+                                         std::chrono::steady_clock::time_point& _time_start) -> void
+    {
+        if (0 == g_pid_ds) {
+            log_server::debug("{}: Checking if Agent Factory is ready to accept client requests before launching the Delay Server.", __func__);
+
+            // Defer the launch of the delay server if the agent factory isn't listening.
+            try {
+                // The host property in server_config.json defines the true identity of the local server.
+                // We cannot use localhost or the loopback address because the computer may have multiple
+                // network interfaces and/or hostnames which map to different IPs.
+                const auto local_server_host = irods::get_server_property<std::string>(irods::KW_CFG_HOST);
+
+                // Use the zone port property from server_config.json. This property defines the port for
+                // server-to-server connections within the zone.
+                const auto local_server_port = irods::get_server_property<int>(irods::KW_CFG_ZONE_PORT);
+                const auto local_server_port_string = std::to_string(local_server_port);
+
+                if (is_server_listening_for_connections(local_server_host, local_server_port_string) != 0) {
+                    return;
+                }
+            }
+            catch (const std::exception& e) {
+                log_server::debug("{}: Connect Error: {}. Deferring launch of Delay Server.", __func__, e.what());
+                return;
+            }
+        }
+
+        const auto migration_sleep_time_in_seconds = irods::get_advanced_setting<int>(irods::KW_CFG_MIGRATE_DELAY_SERVER_SLEEP_TIME_IN_SECONDS);
+        const auto now = std::chrono::steady_clock::now();
+
+        if (now - _time_start < std::chrono::seconds{migration_sleep_time_in_seconds}) {
+            return;
+        }
+
+        _time_start = now;
+
+        const auto hostname = irods::get_server_property<std::string>(irods::KW_CFG_HOST);
+        std::optional<std::string> leader;
+        std::optional<std::string> successor;
+
+        try {
+            irods::experimental::client_connection conn;
+            leader = get_delay_server_leader(conn);
+            successor = get_delay_server_successor(conn);
+
+            if (leader && successor) {
+                log_server::debug("{}: Delay server leader [{}] and successor [{}].", __func__, *leader, *successor);
+
+                // 4 cases:
+                // L  S
+                // ----
+                // 0  0 (invalid state / no-op)
+                // 1  0 (migration complete)
+                // 0  1 (leader has shut down ds, successor launching ds)
+                // 1  1 (leader running ds, admin requested migration)
```

Comment:
consider labeling these cases 0-3 and then referencing these cases in the code below... to make it easier for a later reader to 'see' that all the cases are covered.

Reply:
Done.